mqtt/mqtt-bridge/src/client.rs

```diff
@@ -0,0 +1,227 @@
+#![allow(dead_code)] // TODO remove when ready
+use std::{collections::HashSet, convert::TryInto, time::Duration};
+
+use async_trait::async_trait;
+use chrono::Utc;
+use futures_util::future::BoxFuture;
+use tokio::{net::TcpStream, stream::StreamExt};
+use tracing::{debug, error};
+
+use mqtt3::{
+    proto, Client, Event, IoSource, ShutdownError, SubscriptionUpdateEvent, UpdateSubscriptionError,
+};
+
+use crate::settings::Credentials;
+use crate::token_source::{SasTokenSource, TokenSource};
+
+const DEFAULT_TOKEN_DURATION_MINS: u64 = 60;
+const DEFAULT_TOKEN_DURATION_SECS: Duration = Duration::from_secs(60 * DEFAULT_TOKEN_DURATION_MINS);
+const DEFAULT_MAX_RECONNECT: Duration = Duration::from_secs(5);
+
+#[derive(Debug)]
+pub struct ShutdownHandle(mqtt3::ShutdownHandle);
+
+impl ShutdownHandle {
+    pub async fn shutdown(&mut self) -> Result<(), ClientConnectError> {
+        self.0
+            .shutdown()
+            .await
+            .map_err(ClientConnectError::ShutdownClient)?;
+        Ok(())
+    }
+}
+
+pub struct TcpConnection<T>
+where
+    T: TokenSource + Clone + Send + Sync + 'static,
+{
+    address: String,
+    token_source: Option<T>,
+}
+
+impl<T> IoSource for TcpConnection<T>
+where
+    T: TokenSource + Clone + Send + Sync + 'static,
+{
+    type Io = TcpStream;
+    type Error = std::io::Error;
+    type Future = BoxFuture<'static, Result<(TcpStream, Option<String>), std::io::Error>>;
+
+    fn connect(&mut self) -> Self::Future {
+        let address = self.address.clone();
+        let token_source = self.token_source.as_ref().cloned();
+
+        Box::pin(async move {
+            let expiry = Utc::now()
+                + chrono::Duration::from_std(DEFAULT_TOKEN_DURATION_SECS).unwrap_or_else(|_| {
```

Comment:
Can we make this a const or a static `chrono::Duration` instead of `std::time::Duration` defined on top of the module?

Reply:
I changed it to have the minutes as const and create it form that. I doesn't make sense to create it from string and unwrap. Let me know if there is a better way, as const it didn't compile

Reply:
Ok, I see a complication here...
`std::time::Duration` has `const fn from_secs()` function but `chrono::Duration` does not have that...
So in order to do what I suggested you need to import lazy_static crate and make
```rust
lazy_static! {
    static ref DEFAULT_TOKEN_DURATION: chrono::Duration = chrono::Duration::minutes(60);
}
```
or just leave it with `_MINS` integer.